internal/executor/metrics/metrics.go

```diff
@@ -0,0 +1,81 @@
+// +build !windows !arm
+
+package metrics
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/cirruslabs/cirrus-ci-agent/api"
+	"github.com/shirou/gopsutil/cpu"
+	"github.com/shirou/gopsutil/mem"
+	"time"
+)
+
+var (
+	ErrFailedToQueryCPU    = errors.New("failed to query CPU usage")
+	ErrFailedToQueryMemory = errors.New("failed to query memory usage")
+)
+
+func Run(ctx context.Context) (chan *api.ResourceUtilization, chan error) {
+	resultChan := make(chan *api.ResourceUtilization, 1)
+	errChan := make(chan error, 1)
+
+	go func() {
+		result := &api.ResourceUtilization{}
+
+		pollInterval := 1 * time.Second
+		timeSinceStart := time.Duration(0)
+
+		for {
+			// CPU usage
+			percentages, err := cpu.PercentWithContext(ctx, pollInterval, true)
+			if err != nil {
+				if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+					resultChan <- result
+				} else {
+					errChan <- fmt.Errorf("%w: %v", ErrFailedToQueryCPU, err)
+				}
+				return
+			}
+
+			var numCpusUsed float64
+
+			for _, singleCpuUsageInPercents := range percentages {
+				numCpusUsed += singleCpuUsageInPercents / 100
+			}
+
+			// Memory usage
+			virtualMemoryStat, err := mem.VirtualMemoryWithContext(ctx)
+			if err != nil {
+				if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+					resultChan <- result
+				} else {
+					errChan <- fmt.Errorf("%w: %v", ErrFailedToQueryMemory, err)
+				}
+				return
+			}
+
+			timeSinceStart += pollInterval
```

Comment:
Is it accurate though? Getting the CPU and Memory can take time which won't be accumulated in this mothid. Maybe use `time.Now()` and ceil to the closest second?

Reply:
Fixed in https://github.com/cirruslabs/cirrus-ci-agent/pull/166/commits/21010dfb6c5df0b56b25271b147c0d972dedba6f.